### DISCUSSION TITLE
How to download files after login to site on WPF by CefSharp?

### DISCUSSION THREAD

**Discussion by miladghiravani** [cefsharp/CefSharp#3569] [Q&A] [answered]
Hello

I first enter my account on browser (because without login can't download file.) 
Show browser in my form:

```
browser = new CefSharp.WinForms.ChromiumWebBrowser("https://yoursite.com")
{
   Dock = DockStyle.Fill,
   Size = new Size(600, 600),
   Location = new Point(200, 200),
};
this.panel1.Controls.Add(browser);
```

How to download file by url programmatically by CefSharp and save on pc?

For example like this:

`browser.DownloadFile("https://yoursite.com/download/public/a.zip", "D:\\a.zip");`

Thanks

**Comment by miladghiravani** [comment, marked as answer]
Nobody can help with this question?

**Reply by amaitland** [reply]
> Nobody can help with this question?

Please be patient, no need for a follow up. I will answer questions when time permits.

I see you've also asked at https://stackoverflow.com/questions/67583890/cefsharp-download-files-after-login-to-site-on-wpf

In future please ask your question here **or** stackoverflow and wait for a reply. Not both, thanks. 

> For example like this:
> 
> `browser.DownloadFile("https://yoursite.com/download/public/a.zip", "D:\\a.zip");`

There's no such method.

You can call [IBrowserHost.StartDownload](http://cefsharp.github.io/api/90.6.x/html/M_CefSharp_IBrowserHost_StartDownload.htm) and implement [IDownloadHandler](http://cefsharp.github.io/api/90.6.x/html/T_CefSharp_IDownloadHandler.htm).

Following was written in notepad, so you'll have to fix any minor mistakes
```c#
chromiumWebBrowser.DownloadHandler = new MyDownloadHandler();

public class MyDownloadHandler : IDownloadHandler
{
	public void OnBeforeDownload(IWebBrowser chromiumWebBrowser, IBrowser browser, DownloadItem downloadItem, IBeforeDownloadCallback callback)
	{
		if (!callback.IsDisposed)
		{
			using (callback)
			{
				callback.Continue(System.IO.Path.Combine(@"c:\temp\downloadFolder", downloadItem.SuggestedFileName), showDialog: false);
			}
		}
	}

	public void OnDownloadUpdated(IWebBrowser chromiumWebBrowser, IBrowser browser, DownloadItem downloadItem, IDownloadItemCallback callback)
	{
		
	}
}
```

You can inherit from [DownloadHandler](http://cefsharp.github.io/api/90.6.x/html/T_CefSharp_Handler_DownloadHandler.htm) rather than `IDownloadHandler` and override the methods you require.

----

Alternatively you can use [IFrame.CreateUrlRequest](http://cefsharp.github.io/api/90.6.x/html/M_CefSharp_IFrame_CreateUrlRequest.htm) to manually construct a request and save the download yourself to disk.

https://github.com/cefsharp/CefSharp/blob/cefsharp/90/CefSharp.Test/OffScreen/OffScreenBrowserBasicFacts.cs#L292 should be a base that you can adapt to implement this yourself. You'll need https://github.com/cefsharp/CefSharp/blob/4608924d5a0c676857720917550674b88bd7b701/CefSharp.Example/UrlRequestClient.cs#L10 as well.

**Reply by miladghiravani** [reply]
@amaitland Thank you, Yes you are right.
Thank you for your best answer
